### ISSUE TITLE
Bug on OS X, due to linker's incomplete support of thread local storage class

### ISSUE THREAD

**Issue by thewhimer** [feeley/gambit#109]
I tried to use Gambit-C with the 'multiple-vms’ option enabled, to harness full cpu power by SMP. But ‘compile-file’ on REPL failed with the message:

ld: illegal thread local variable reference to regular symbol ____tls_ptr for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

After some experiments, it seems that the OS X linker does not properly support dynamic linking of externally defined TLS variables. This can be verified easily with the following code.

// test.c
extern __thread int tls_id;
int get_tls_id()
{
    return tls_id;
}

On OS X, both gcc and clang failed to produce a dynamically loadable object file.
(OS X 10.10.1, Xcode 6.2, clang-600.0.57; gcc-4.8 intalled from homebrew.)
$ gcc-4.8 -shared test.c -o libfoo.so
$ gcc-4.9 -shared test.c -o libfoo.so
$ clang -bundle test.c -o libfoo.so

These three give the same error:

Undefined symbols for architecture x86_64:
  "_tls_id", referenced from:
      _get_tls_id in test-9d6352.o
     (maybe you meant: _get_tls_id)
ld: symbol(s) not found for architecture x86_64

While on Linux (ArchLinux 3.18.6 X86_64, gcc 4.9.2), it worked. The produced library linked fine with main.c (as follows), execution gives 1.

// main.c
#include <stdio.h>

__thread int tls_id;

extern int get_tls_id();

int main ()
{
    int id = 0;
    tls_id = 1;
    id = get_tls_id();
    printf(“tls id is %d\n”, id);
    return 0;
}

It might be tested in configure script, and should fallback to pthread_get/setspecific (POSIX) if not supported.

**Comment by feeley** [comment]
That's a bummer!  pthread_getspecific is much less efficient...  I wonder if Apple is working on a fix.

**Comment by jncornett** [comment]
This is actually not limited to OS X. LLVM doesn't support extern __thread variables. You can, however use C++11 thread_local type variables just fine. I filed a bug with LLVM in December, but I don't think anyone has looked at it for whatever reason: https://llvm.org/bugs/show_bug.cgi?id=25821

**Comment by thewhimer** [comment]
This should be easy to fix, a helper function could be used to retrieve the external TLS variable for processor state, with minimal performance cost.

Actually, this helper function already exist, ```___get_tls_ptr```, currently used when TLS is not supported. Unified access through ```___get_tls_ptr``` is sensible, whether TLS is supported or not.

**Comment by feeley** [comment]
I wonder if it is possible on OS X to have a custom allocation function for pthread_t .  If that's the case, then a block of memory the size of struct { pthread_t pt; void *tls_ptr; } could be allocated and it would be possible to get to tls_ptr from pthread_self.

**Comment by feeley** [comment]
I found that NaCL has an interesting implementation of TLS on OS X.  Look at function NaClTlsInit in https://src.chromium.org/native_client/trunk/src/native_client/src/trusted/service_runtime/arch/x86_64/nacl_tls_64.c .  This might be a source of inspiration to implement efficient TLS on OS X.  It could also mean that pthread_getspecific is already fairly efficient on OS X (basically indexing an array).  This should be tested.

**Comment by thewhimer** [comment]
OS X does support TLS.
It just does not support referencing an externally defined TLS variable during link time. No support is actually a good thing, directly referencing an external TLS variable should be a bad style.

**Comment by thewhimer** [comment]
@feeley I have submitted a pull request, fixing this issue as described above.

To avoid the getter overhead in the gambit library, ___GET_PSTATE() might be defined accordingly only for the compilation of gambit library. But I have not found a macro which discriminates internal (the library) and external module compilation.

**Comment by vyzo** [comment]
The problem is that it's in the fastpath and that isn't going to fly past marc; twice as slow on something there makes for measurable performance variations. 
Why don't you simply make it a macro that expands to a function call on systems that need it and a direct access otherwise?

**Comment by thewhimer** [comment]
@feeley Thanks for the fix [dbbb8f5](https://github.com/feeley/gambit/commit/dbbb8f58dcd89a76e17e83d74672a59fe41ed4af). But a minor problem, make check still fails. I had to comment out #ifndef ___PRIMAL as below.

```
#ifndef ___DYNAMIC_TLS

#undef ___DEFINE_THREAD_LOCAL_STORAGE_GETTER_SETTER
#define ___DEFINE_THREAD_LOCAL_STORAGE_GETTER_SETTER

#ifdef ___SHARED
// #ifndef ___PRIMAL
#undef ___USE_THREAD_LOCAL_STORAGE_GETTER_SETTER
#define ___USE_THREAD_LOCAL_STORAGE_GETTER_SETTER
// #endif
#endif

#endif

#else
```

This is because ```___DYNAMIC``` always implies ```___PRIMAL```, as in gambit.h.in
```
/*
 * Define symbols appropriate for dynamic loading.
 */

#ifdef ___DYNAMIC
#undef ___PRIMAL
#define ___PRIMAL
#undef ___LIBRARY
#define ___LIBRARY
#undef ___SHARED
#define ___SHARED
#undef ___BIND_LATE
#define ___BIND_LATE
#endif
```

And as ```___PRIMAL``` is already define in ```lib/makefine.in``` for compiling the core library. It should be removed from where as quoted above. After that, this issue could be closed.

Sorry for this pet peeve.
